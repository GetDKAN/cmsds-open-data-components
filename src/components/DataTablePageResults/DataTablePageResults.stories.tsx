import React from 'react';
import DataTablePageResults from './DataTablePageResults';

const meta = {
  title: 'Components/DataTablePageResults',
  component: DataTablePageResults,
  parameters: {
    layout: 'padded',
    docs: {
      description: {
        component: `\nThe DataTablePageResults component displays the current range of visible rows and the total number of results in a data table.\n        `,
      },
    },
  },
  argTypes: {
    totalRows: { control: 'number', description: 'Total number of rows.' },
    limit: { control: 'number', description: 'Rows per page.' },
    offset: { control: 'number', description: 'Offset for current page.' },
    className: { control: 'text', description: 'Custom class name.' },
  },
  tags: ['autodocs'],
};

export default meta;

export const Default = {
  args: {
    totalRows: 100,
    limit: 25,
    offset: 0,
    className: 'data-table-results',
  },
};

export const MiddlePage = {
  args: {
    totalRows: 100,
    limit: 25,
    offset: 25,
    className: 'data-table-results',
  },
};

export const LastPage = {
  args: {
    totalRows: 100,
    limit: 25,
    offset: 75,
    className: 'data-table-results',
  },
};

export const NoResults = {
  args: {
    totalRows: 0,
    limit: 25,
    offset: 0,
    className: 'data-table-results',
  },
};
