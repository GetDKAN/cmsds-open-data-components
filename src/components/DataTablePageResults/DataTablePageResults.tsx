import React from 'react';
import PropTypes from 'prop-types';

export type DataTablePageResultsType = {
  totalRows: number;
  limit: number;
  offset: number;
  className?: string;
}

const DataTablePageResults = ({
  totalRows, limit, offset, className = 'data-table-results'
}: DataTablePageResultsType) => {
  const numTotalRows = totalRows;
  if (numTotalRows === 0) {
    return <p className={className}>{`0 - 0 of 0 rows`}</p>
  }
  const ofTotal = () => {
    if (limit >= numTotalRows) { return numTotalRows; }
    if (limit + offset >= numTotalRows) { return numTotalRows; }
    if (offset === 0) { return limit; }
    return (offset + limit);
  }
  const page = offset / limit;
  const startTotal = () => (page * limit + 1)
  return (
    <p className={className}>
      Displaying{' '}
      <span className="ds-u-font-weight--bold">{`${startTotal().toLocaleString()} - ${ofTotal().toLocaleString()}`}</span>{' '}
      of <span className="ds-u-font-weight--bold">{`${numTotalRows.toLocaleString()}`}</span>{' '}
      rows
    </p>
  );
};

DataTablePageResults.propTypes = {
  className: PropTypes.string,
  totalRows: PropTypes.number.isRequired,
  limit: PropTypes.number.isRequired,
  offset: PropTypes.number.isRequired,
};

export default DataTablePageResults;