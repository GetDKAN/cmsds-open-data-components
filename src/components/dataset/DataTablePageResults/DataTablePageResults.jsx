import React from 'react'
import PropTypes from 'prop-types'

const DataTablePageResults = ({
  total,
  pageSize,
  currentPage,
  className = 'data-table-results',
  viewing = false
}) => {
  // Add one to offset the 0 array index.
  const safeCurrentPage = Number.isFinite(currentPage) ? currentPage : 0
  const safePageSize = Number.isFinite(pageSize) ? pageSize : 0
  const page = safeCurrentPage + 1
  let displayTotal = total

  const currentLowestResult = total <= 0 ? 0 : 1 + ((safePageSize * page) - safePageSize)
  let currentHighestResult = (safePageSize * page)
  if (total < 0) {
    displayTotal = 0
  }
  if (currentHighestResult > total) {
    currentHighestResult = displayTotal
  }
  return (
    <div className={className}>
      <p>
        {viewing && (
          'Viewing '
        )}
        <span className="low-result">{currentLowestResult}</span>
        {' '}
        -
        {' '}
        <span className="high-result">{currentHighestResult}</span>
        {' '}
        of
        {' '}
        <span className="total">{displayTotal.toLocaleString()}</span>
        {' '}
        rows
      </p>
    </div>
  )
}

DataTablePageResults.propTypes = {
  className: PropTypes.string,
  total: PropTypes.number.isRequired,
  pageSize: PropTypes.number.isRequired,
  currentPage: PropTypes.number.isRequired,
  viewing: PropTypes.bool
}

export default DataTablePageResults
