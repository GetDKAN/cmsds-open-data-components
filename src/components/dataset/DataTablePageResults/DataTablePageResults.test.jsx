import { render, fireEvent } from '@testing-library/react'
import DataTablePageResults from './DataTablePageResults'

const componentArgs = {
  total: 100,
  pageSize: 20,
  currentPage: 0,
  className: '',
  viewing: false
}

const renderComponent = (args = componentArgs) => render(
  <DataTablePageResults {...args} />
)

describe('DataTablePageResults component.', () => {
  it('Matches snapshot.', () => {
    const renderedDataTablePageResults = renderComponent({
      ...componentArgs,
      viewing: undefined
    })

    expect(renderedDataTablePageResults.asFragment()).toMatchSnapshot()
  })

  it('Displays \'Viewing\' text when \'viewing\' prop is true.', () => {
    const { container } = renderComponent({
      ...componentArgs,
      viewing: true,
      className: 'test-container'
    })

    expect(container.querySelector('.test-container').textContent).toBe('Viewing 1 - 20 of 100 rows')
  })

  it('Display total is 0 by default.', () => {
    const { container } = renderComponent({
      ...componentArgs,
      className: 'test-container',
      total: -1
    })

    expect(container.querySelector('.test-container').textContent).toBe('0 - 0 of 0 rows')
  })
})
