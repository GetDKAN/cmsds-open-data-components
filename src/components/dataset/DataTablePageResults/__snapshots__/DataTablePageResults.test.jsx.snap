// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DataTablePageResults component. > Matches snapshot. 1`] = `
<DocumentFragment>
  <div
    class=""
  >
    <p>
      <span
        class="low-result"
      >
        1
      </span>
       - 
      <span
        class="high-result"
      >
        20
      </span>
       of 
      <span
        class="total"
      >
        100
      </span>
       rows
    </p>
  </div>
</DocumentFragment>
`;
